control/discovery.py

```diff
@@ -0,0 +1,1078 @@
+        # admin max SQ size
+        ("asqsz", c_uint16),
+        ("reserved1", c_ubyte * 22),
+        ("trsvcid", c_ubyte * 32),
+        ("reserved2", c_ubyte * 192),
+        ("subnqn", c_ubyte * 256),
+        ("traddr", c_ubyte * 256),
+        # Transport specific address subtype
+        ("tsas", c_ubyte * 256)
+    ]
+
+class DiscoveryService:
+    """Implements discovery controller.
+
+    Response discover request from initiator.
+
+    Instance attributes:
+        version: Discovery controller version
+        config: Basic gateway parameters
+        logger: Logger instance to track discovery controller events
+        omap_name: OMAP object name
+        ioctx: I/O context which allows OMAP access
+        discovery_addr: Discovery controller addr which allows initiator send command
+        discovery_port: Discovery controller's listening port
+    """
+
+    BDEV_PREFIX = "bdev_"
+    NAMESPACE_PREFIX = "namespace_"
+    SUBSYSTEM_PREFIX = "subsystem_"
+    HOST_PREFIX = "host_"
+    LISTENER_PREFIX = "listener_"
+
+    def __init__(self, config):
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(level=logging.DEBUG)
+
+        self.version = 1
+        self.config = config
+        self.lock = threading.Lock()
+
+        gateway_group = self.config.get("gateway", "group")
+        self.omap_name = f"nvmeof.{gateway_group}.state" \
+            if gateway_group else "nvmeof.state"
+        self.logger.info(f"log pages info from omap: {self.omap_name}")
+
+        ceph_pool = self.config.get("ceph", "pool")
+        ceph_conf = self.config.get("ceph", "config_file")
+        conn = rados.Rados(conffile=ceph_conf)
+        conn.connect()
+        self.ioctx = conn.open_ioctx(ceph_pool)
+
+        self.discovery_addr = self.config.get("discovery", "addr")
+        self.discovery_port = self.config.get("discovery", "port")
+        if self.discovery_addr == '' or self.discovery_port == '':
+            self.logger.error("discovery addr/port are empty.")
+            assert 0
+        self.logger.info(f"discovery addr: {self.discovery_addr} port: {self.discovery_port}")
+
+        self.conn_vals = {}
+        self.connection_counter = 1
+        self.selector = selectors.DefaultSelector()
+
+    def _read_all(self) -> Dict[str, str]:
+        """Reads OMAP and returns dict of all keys and values."""
+
+        with rados.ReadOpCtx() as read_op:
+            iter, _ = self.ioctx.get_omap_vals(read_op, "", "", -1)
+            self.ioctx.operate_read_op(read_op, self.omap_name)
+            omap_dict = dict(iter)
+        return omap_dict
+
+    def _get_vals(self, omap_dict, prefix):
+        """Read values from the OMAP dict."""
+
+        vals = []
+        for (key, val) in omap_dict.items():
+            if key.startswith(prefix):
+                val_text = val.decode('utf-8')
+                js = json.loads(val_text)
+                vals.append(js)
+        return vals
```

Comment:
Not a big deal, but there's an expression (list comprehension), which is considered the Pythonic way to iterate and filter through lists/dicts:
```suggestion
        return [json.loads(val.decode('utf-8')) for (key, val) in omap_dict.items()
            if key.startswith(prefix)]
```